src/main/java/cz/muni/fi/pa165/bluebat/entity/Customer.java

```diff
@@ -1,5 +1,95 @@
-package cz.muni.fi.pa165.bluebat.entity;
+package cz.muni.fi.pa165.bluebat.entity;/* created by rudolf */
 
+
+import javax.persistence.*;
+import javax.validation.constraints.NotNull;
+import java.time.LocalDate;
+import java.util.Objects;
+
+/**
+ * @author rudolf
+ */
+@Entity
 public class Customer {
+    public Customer() {
+
+    }
+
+    public Customer(Long number) {
+
+        this.phoneNumber = number;
+    }
+
+    @OneToOne
+    private Reservation reservation;
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE)
+    private Long id;
+
+    private LocalDate birthday;
+
+
+    private String name;
+
+    private String surname;
+
+    public String getSurname() {
+        return surname;
+    }
+
+    public void setSurname(String surname) {
+        this.surname = surname;
+    }
+
+    @NotNull
+    @Column(nullable = false, unique = true)
+    private Long phoneNumber;
```

Comment:
Nevím jestli je dobrý nápad mít telefonní číslo unikátní. Pokud správně chápu náš návrh aplikace, tak pro každou rezervaci by se měl vytvořit nový zákazník, který by obsahoval data vyplněná uživatelem při vytváření rezervace (podobné, když si objednávám v e-shopu a vyplňuji objednávku). Tak bych chtěl mít možnost vytvořit více rezervací i se stejným telefonním číslem. Kdyby se uživatel musel nejprve registrovat, tak by to smysl dávalo.

Reply:
Chapem,, ale tak potom musime mat nkatne id co nie je velmi dobre.

Reply:
*unikátne ID

Reply:
A potom je problem v hash a equals, to nam hovoril na cviku aj v prednaske ze id by sa nemalo nikdy davat do quals a hashu.

Reply:
Je to takové blbé, ale zase nevidím rozumný způsob, jak vytvořit další rezervaci se stejným číslem. Jedině, kdyby zadal telefonní číslo, my zjistili, že už rezervaci vytvářel a nechtít po něm vyplnění dalších údajů 🤔 ale i je to taky trochu nepraktické

Reply:
To máš pravdu, že to id by nemělo být v equals ani hashu 🤔

Reply:
To sa mi nezda, mozme mat uplene toho isteho cloveka 100 krat v databaze ako to budeme nachadzat, to bude zmetok. Ja to vidim tak ze ich registrovat alebo nejake unikatne nieco ako cislo zakazky.

Reply:
No pokud si to bude ukládat rezervace jako @JoinColumn nebo @JoinTable, tak si vždy uchová id zákazníka, který je s danou rezervací spojený

Reply:
Zajímavý problém. Jako nepřijde mi zase tak problematické se podívat, zda už daného zákazníka nemáme v databázi (ať už podle tel. čísla nebo podle emailu) a podle toho se pak zachovat. Ale fakt nevím, který z těch dvou přístupů je lepší

Reply:
Já jsem naopak pro duplicitní zákazníky...podle mě je z analytického pohledu správnou možností, jak pracovat s již existujícím zákazníkem, nějaká forma autentizace,  do čehož se asi úplně pouštět nechceme.

V případě, že bychom se rozhodli pro využití již stávajících dat, byl bych určitě pro, aby se tam doplnil nějaký jedinečný atribut (např. RČ) či se alespoň zákazník vyhledával na základě kombinace všech existujících atributů. Tel. číslo ani mail nelze úplně považovat za nějakou jedinečnou informaci.

Reply:
Samozřejmě, že nejvhodnějším řešením by bylo vyřešit autentizaci. Ale musíme najít nějakou nejméně bolestivou alternativu. Vylistování všech unikátních zákazníků by asi pořád nebyl problém, jelikož by to mělo stačil hodit do Setu a tím, že duplicity mají stejné equals, by se tam měla dostat jenom jedna. Pokud bychom chtěli vyhledat zákazníka třeba podle jména, zase se to komplikuje tím, že bychom jich našli X a pak bychom ještě museli rozlišovat podle konkrétní rezervace. Samozřejmě by pak nebylo jednoduché modifikovat nějaký údaj přes všechny rezervace zákazníka, ale to problém není, to prostě němá cenu řešit, když neděláme tu autentizaci.

Pokud bychom přistoupili k řešení, kde máme na jednoho zákazníka list rezervací, tak se to zase komplikuje na začátku při tom vytváření rezervace, kde musíme zjistit, zda už takového zákazníka nemáme. Každopádně podle mailu mi to přijde v pohodě, pokud bychom dělali autentizaci, taky budeme uživatele rozlišovat podle emailu, který by zároveň mohl sloužit jako login.  To že jeden člověk může použít dva různé emaily a založit si dva účty už pak podle mě není něco, co by nás muselo zajímat.

Každopádně čím víc nad tím přemýslím, tím víc se začínám klanět k variantě, kde budeme mít zákazníky a rezervace 1:1

Reply:
Takže pridám adresu a email. Email dám unikát. Equals a hash pomocou emailu.Vymažem vyhldavanie pomocou mena a telefónneho cisla. Pridám vyhladavanie pomocou emailu. Pridám zoznam vsetkych customers (findAll). OK ?

Reply:
Tak teoreticky i pod mailem se může skrývat více lidí, jejichž jediným společným atributem je právě ten mail (např. První rezervaci děláš sám za sebe, druhou rezervaci děláš babičce, která nemá mail, takže tam vyplníš svůj mail.). Toto by se dalo asi vyřešit tím, že se ti to automaticky předvyplní na základě mailu a zakáže se další editace, což ale s sebou přináší další problémy.

Určitě souhlasím, že bychom měli zvolit to nejméně bolestivé řešení. Uvádím zde už takové ty "hnidopišské připomínky". Asi hodně záleží, jestli to bude hodnotit po té programátorské stránce či po stránce uživatelské. Na trh s tím asi jít neplánujeme :D

Jak jsem psal, taky jsem pro variantu 1:1. Takže ještě uvidíme, jaký názor bude mít zbytek.

Reply:
No pokud chceme mít zákazníky a rezervace 1:1 tak by ani email neměl být unikátní

Reply:
OK, jedne jedinecne co tam je je tak rezervacia to mozem porovnavat ale v Reservation je equas a hash urobeny porvnavani customer a Trip. Alebo to mozmer robit ze hodim vsetky atributy do eqals a hash. Tu je problem ze v class diagramme mame 1:1 Trip, Reservation, Customer co je vlasne jedna classa rozdelena do troch.